src/osd/OSD.cc

```diff
@@ -3536,13 +3536,7 @@ int OSD::update_crush_device_class()
     string("\"class\": \"") + device_class + string("\", ") +
     string("\"ids\": [\"") + stringify(whoami) + string("\"]}");
 
-  r = mon_cmd_maybe_osd_create(cmd);
-  // the above cmd can fail for various reasons, e.g.:
-  //   (1) we are connecting to a pre-luminous monitor
-  //   (2) user manually specify a class other than
-  //       'ceph-disk prepare --crush-device-class'
-  // simply skip result-checking for now
-  return 0;
+  return mon_cmd_maybe_osd_create(cmd);
```

Comment:
As per discussion with Sage, we generally do not allow silently changing the device class of an OSD if it already has been bound to a specific one, so there is still a special case we want to tolerate the error returned by mon:
(1) The OSDs newly added into the cluster will be automatically assigned a class type - hdd/ssd/nvme, based on its backend media type.
(2) User manually change their device classes to any freeform(e.g., "ssd-for-db", "hdd-for-backup") through CLI.
(3) OSD will now refuse to start up after reboot since it will try to reset its device-class  to hdd/sdd/nvme which is not allowed by Mon

Reply:
i think in that case user is required to remove the device class first. because:

> we generally do not allow silently changing the device class of an OSD if it already has been bound to a specific one

am i understanding this correctly?

Reply:
> i think in that case user is required to remove the device class first

Generally user is not aware of that all OSDs are automatically assigned a default class unless we turn osd_class_update_on_start off. So:
(1) new osds are automatically labelled as per their backend media type, say, "hdd"
(2) user does "ceph osd crush rm-device-class all", and 
     "ceph osd crush set-device-class abc osd.1 osd.2"
     "ceph osd crush set-device-class def osd.3 osd.4"
     ... to change all osds's device class as per their preference
(3) restart osds, and if osd_class_update_on_start is still on, osds will try to repeat step (1)(
     "ceph osd crush set-device class hdd osd.*"), which shall fail...

Of course the above problem can be eliminated by manually shutting osd_class_update_on_start down after step(2). But for now I think we can simply leave this as it is and fix the warning instead.

What do you think?

Reply:
that's embarrassing.. can we `ceph osd crush rm-device-class osd.x` when osd.x is up and running?

Reply:
Sorry, kefu, I think we should respect user's choice. 
If we silently do rm-device-class and  then automatically alter its device class to a different kind, it might cause unexpected (even wrong) behaviour.

Reply:
cool, so i will amend the commit, so we will return the error if it is not `-EBUSY`. how about this?

Reply:
also -EPERM (normal during upgrade)

Reply:
@liewegas we've dropped the code handling pre-luminous releases. so "osd crush set-device-class" won't return `-EPERM`, i think.

Reply:
That's right, we are in mimic now:-)